controllers/kubeadmconfig_controller.go

```diff
@@ -394,6 +384,46 @@ func (r *KubeadmConfigReconciler) reconcileDiscovery(cluster *capiv1alpha2.Clust
 	return nil
 }
 
+// reconcileTopLevelObjectSettings injects into config.ClusterConfiguration values from top level objects like cluster and machine.
+// The implementation func respect user provided config values, but in case some of them are missing, values from top level objects are used.
+func (r *KubeadmConfigReconciler) reconcileTopLevelObjectSettings(cluster *capiv1alpha2.Cluster, machine *capiv1alpha2.Machine, config *cabpkv1alpha2.KubeadmConfig) {
+	log := r.Log.WithValues("kubeadmconfig", fmt.Sprintf("%s/%s", config.Namespace, config.Name))
+
+	// If there are no ControlPlaneEndpoint defined in ClusterConfiguration but there are APIEndpoints defined at cluster level (e.g. the load balancer endpoint),
+	// then use cluster APIEndpoints as a control plane endpoint for the K8s cluster
+	if config.Spec.ClusterConfiguration.ControlPlaneEndpoint == "" && len(cluster.Status.APIEndpoints) > 0 {
+		// NB. CABPK only uses the first APIServerEndpoint defined in cluster status if there are multiple defined.
+		config.Spec.ClusterConfiguration.ControlPlaneEndpoint = fmt.Sprintf("%s:%d", cluster.Status.APIEndpoints[0].Host, cluster.Status.APIEndpoints[0].Port)
+		log.Info("Altering ClusterConfiguration", "ControlPlaneEndpoint", config.Spec.ClusterConfiguration.ControlPlaneEndpoint)
+	}
+
+	// If there are no ClusterName defined in ClusterConfiguration, use Cluster.Name
+	if config.Spec.ClusterConfiguration.ClusterName == "" {
+		config.Spec.ClusterConfiguration.ClusterName = cluster.Name
+		log.Info("Altering ClusterConfiguration", "ClusterName", config.Spec.ClusterConfiguration.ClusterName)
+	}
+
+	// If there are no Newtwork settings defined in ClusterConfiguration, use ClusterNetwork settings, if defined
+	if config.Spec.ClusterConfiguration.Networking.DNSDomain == "" && cluster.Spec.ClusterNetwork != nil && cluster.Spec.ClusterNetwork.ServiceDomain != "" {
+		config.Spec.ClusterConfiguration.Networking.DNSDomain = cluster.Spec.ClusterNetwork.ServiceDomain
+		log.Info("Altering ClusterConfiguration", "DNSDomain", config.Spec.ClusterConfiguration.Networking.DNSDomain)
+	}
+	if config.Spec.ClusterConfiguration.Networking.ServiceSubnet == "" && cluster.Spec.ClusterNetwork != nil && len(cluster.Spec.ClusterNetwork.Services.CIDRBlocks) > 0 {
+		config.Spec.ClusterConfiguration.Networking.ServiceSubnet = strings.Join(cluster.Spec.ClusterNetwork.Services.CIDRBlocks, "")
+		log.Info("Altering ClusterConfiguration", "ServiceSubnet", config.Spec.ClusterConfiguration.Networking.ServiceSubnet)
+	}
+	if config.Spec.ClusterConfiguration.Networking.PodSubnet == "" && cluster.Spec.ClusterNetwork != nil && len(cluster.Spec.ClusterNetwork.Pods.CIDRBlocks) > 0 {
+		config.Spec.ClusterConfiguration.Networking.PodSubnet = strings.Join(cluster.Spec.ClusterNetwork.Pods.CIDRBlocks, "")
+		log.Info("Altering ClusterConfiguration", "PodSubnet", config.Spec.ClusterConfiguration.Networking.PodSubnet)
+	}
```

Comment:
We could wrap these past 3 `if` conditions in a block starting with `cluster.Spec.ClusterNetwork != nil`